### ISSUE TITLE
How long should training take?

### ISSUE THREAD

**Issue by ArnaudDeza** [grimmlab/gumbeldore#2]
Hi,

How long should training take for the experiments in the paper? 

I am trying to train the TSP model from scratch with the hardware setup described in the appendix (16 workers and 2 GPUs) however one epoch is roughly 1-1.5 hours.  Is this a normal amount of time when using the default config for TSP or am I doing something wrong with the ray parallelization?

P.S: This paper is truly amazing and very well written. I appreciate that you have made the code public for the broader research community!

Arnaud

**Comment by johnnycrab** [comment]
Hi Arnaud,

Thanks for reaching out! How long one epoch takes depends of course on your CPU and GPU, but with 16 workers und 2 GPUs, 1-1.5 hours is definitely way too long. I am certain that my code could be improved and sped up 😅 , but we should try to push it down to at most 10 minutes. Can you send me the config file that you are using, and your hardware setup? Also, can you confirm with `$ watch nvidia-smi` that the GPUs are used? Depending on your hardware, it can also help to set `pin_workers_to_core` to `True` in the `gumbeldore_config` dictionary in the config. This pins the ray processes to individual threads, because it can happen that if you have multiple processes running numpy/pytorch, then they might slow each other down. This is kind of a dirty workaround, but I found that this helps in case you have on your CPU, e.g., 16 threads and use exactly 16 ray workers. 
We'll figure it out. :) 

PS: Thanks for your kind words! We also have a simpler sequence decoding technique that can be plugged in here: https://github.com/grimmlab/step-and-reconsider . It uses the same code base, but has a different way of dealing with the sampled sequences in each round, that doesn't require computing an advantage and scaling it. Just if you're interested. :)

**Comment by ArnaudDeza** [comment]
Hi,

Sorry for the late reply and thanks for your quick response. I realized I was not setting "devices_for_workers" and "CUDA_VISIBLE_DEVICES" appropriately in the config file.

With 32 cores and 5 GPU's one epoch for the TSP dataset takes 3-5 minutes (training + validation) and all cores and GPUs are at 100% utilization.

Sorry for not checking this beforehand and thank you for your help!

**Comment by johnnycrab** [comment]
No need to apologize, I'm relieved that it works! Let me know if you run into other problems! :)
